src/flow-control/base-handler.js

```diff
@@ -0,0 +1,16 @@
+module.exports = class BaseHandler {
+    constructor(options = {}) {
+        const { logger } = options;
+
+        this.logger = logger;
+    }
+
+    handle(commands) {
+        return commands;
+    }
+
+    /**
+     * A hook called when all commands have finished (either successful or not).
+     */
+    onFinish() {}
```

Comment:
WDYT if `handle` returned an optional cleanup function, along with the mapped commands?
Having an `onFinish` method kinda implies that the instance is single use.

Reply:
Personally, I like this better:
* It keeps `handle` as a pure transformation `Commands => Commands`
* It makes the `reduce` logic a bit more complicated in the main function
* It might also be a breaking change, if any third-party APIs write their own handlers
    * The only way to not make it a breaking change would be to allow `handle` to return either `Commands` or `{ commands, cleanup }`, which is additional complexity for the caller to handle

But it's your call

Reply:
For this one I think it's better to introduce a breaking change.
As you correctly stated, third parties can pass their own handlers - and it could become a pain point to manage state across multiple runs of the same instance.

Some other uses I can see for this cleanup function:
- Removing the RxJS subscription (which also removes the event from the `inputStream`, though doesn't make the process exit if it's `stdin`)
- Removing listeners on signals in the `KillOnSignal` handler (which can also prevent the process from exiting)

Reply:
Yeah, that makes sense. I can update the handle api if you want. Are there any docs that I'd need to update as well?

Reply:
Nope, no docs on creating custom handlers to this date 😞